### PR TITLE
An algorithm for finding the nearest neighbour

### PR DESCRIPTION
**Pull request by pleroy**
Related to #3358.

### CODE REVIEW

File: numerics/nearest_neighbour_body.hpp [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    if (projection < Norm{}) {
+      other_side = internal.children.second.get();
+    } else {
+      other_side = internal.children.first.get();
+    }
+
+    // We omit |must_check_other_side| because there is no point in checking the
+    // other side again.
+    // �Maintenant, rien qu'en traversant, j'ai fait qu'en face soit en face.
+    // L'autre c�t� a chang� de c�t�!� [GT71]
```
Comment (eggrobin) [inline numerics/nearest_neighbour_body.hpp:212]: Spaces (non-breaking) after « and ».

File: numerics/nearest_neighbour.hpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+// Principal component partitioning trees (PCP trees) are introduced by [WZ91]
+// in the context of quantization.  Their use for nearest neighbour search was
+// proposed by [ZJL02].  We choose to use this algorithm not because it is easy
+// but because it is coordinate-free.
+template<typename Value_>
+class PrincipalComponentPartitioningTree {
+ public:
+  using Value = Value_;
+
+  // We stop subdividing a cell when it contains |max_values_per_cell| or less
```
Comment (eggrobin) [inline numerics/nearest_neighbour.hpp:35]: fewer.

File: numerics/nearest_neighbour_body.hpp [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+          values,
+          std::vector<double>(values.size(), 1))) {
+  displacements_.reserve(values.size());
+  for (auto const& value : values) {
+    displacements_.push_back(value - centroid_);
+  }
+
+  Indices indices;
+  indices.reserve(displacements_.size());
+  for (int i = 0; i < displacements_.size(); ++i) {
```
Comment (eggrobin) [inline numerics/nearest_neighbour_body.hpp:41]: Should we check that `displacements_` isn’t too big here?

**Review by eggrobin — APPROVED** [review]

**Comment by pleroy** [comment]
retest this please

**Pull request merged**